File: services/ark-dashboard/ark-dashboard/lib/services/api-keys.ts
/**
 * API Key management service for ARK Dashboard
 */

import { apiClient } from '../api/client';
import type { components } from '../api/generated/types';

// Use generated types from OpenAPI
export type APIKey = components['schemas']['APIKeyResponse'];
export type APIKeyCreateRequest = components['schemas']['APIKeyCreateRequest'];
export type APIKeyCreateResponse = components['schemas']['APIKeyCreateResponse'];
export type APIKeyListResponse = components['schemas']['APIKeyListResponse'];

// API Key service following ARK Dashboard patterns
export const apiKeysService = {
  /**
   * Get all API keys
   */
  async getAll(): Promise<APIKeyListResponse> {
    return apiClient.get<APIKeyListResponse>('/api/v1/api-keys');
  },

  /**
   * Create a new API key
   */
  async create(request: APIKeyCreateRequest): Promise<APIKeyCreateResponse> {
    return apiClient.post<APIKeyCreateResponse>('/api/v1/api-keys', request);
  },

  /**
   * Delete (revoke) an API key by public key
   */
  async delete(publicKey: string): Promise<void> {
    return apiClient.delete(`/api/v1/api-keys/${publicKey}`);
  }
};

File: services/ark-api/ark-api/src/ark_api/api/v1/api_keys.py
"""API key management endpoints."""

import logging
from typing import List
from fastapi import APIRouter, HTTPException, status

from ...models.auth import (
    APIKeyCreateRequest,
    APIKeyResponse,
    APIKeyCreateResponse,
    APIKeyListResponse
)
from ...services.api_keys import APIKeyService
from .exceptions import handle_k8s_errors

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api-keys", tags=["api-keys"])


@router.post("", response_model=APIKeyCreateResponse, status_code=status.HTTP_201_CREATED)
async def create_api_key(
    body: APIKeyCreateRequest,
) -> APIKeyCreateResponse:
    """
    Create a new API key for service-to-service authentication.
    
    Args:
        body: API key creation request
        
    Returns:
        APIKeyCreateResponse: The created API key with secret (only shown once)
    """
    try:
        api_key_service = APIKeyService()
        
        # No user attribution in current global key model - see docs/basic-auth-future-enhancements.md
        user_id = None
        
        result = await api_key_service.create_api_key(body, user_id=user_id)
        
        logger.info(f"Created API key '{body.name}' with public key {result.public_key}")
        return result
        
    except Exception as e:
        logger.error(f"Error creating API key: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to create API key: {str(e)}"
        )


@router.get("", response_model=APIKeyListResponse)
async def list_api_keys() -> APIKeyListResponse:
    """
    List all active API keys (without secret keys).
    
    Returns:
        APIKeyListResponse: List of API keys
    """
    try:
        api_key_service = APIKeyService()
        result = await api_key_service.list_api_keys()
        
        logger.debug(f"Listed {result.count} API keys")
        return result
        
    except Exception as e:
        logger.error(f"Error listing API keys: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to list API keys: {str(e)}"
        )


@router.delete("/{public_key}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_api_key(public_key: str):
    """
    Soft delete an API key by marking it as inactive.
    
    Args:
        public_key: The public key of the API key to delete
    """
    try:
        api_key_service = APIKeyService()
        
        success = await api_key_service.delete_api_key(public_key)
        if not success:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"API key with public key '{public_key}' not found"
            )
        
        logger.info(f"Deleted API key {public_key}")
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error deleting API key {public_key}: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to delete API key: {str(e)}"
        )
